### PR TITLE
Basic vault

### PR DESCRIPTION
**Pull request by davidbrai**
Implementation of a basic vault which holds an ERC20 token for users.

Assignment details: https://github.com/yieldprotocol/mentorship2022/issues/2

### CODE REVIEW

File: src/assignment-2/BasicVault.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    /// @param _token The address of an ERC20 token contract
+    constructor(IERC20 _token) {
+        token = _token;
+    }
+
+    /// @notice Deposits tokens from the user into the vault
+    /// @dev Updates the balances mapping with the amount of tokens
+    /// @param amount The amount of tokens to deposit
+    function deposit(uint amount) public {
+        bool success = token.transferFrom(msg.sender, address(this), amount);
```
Comment (devtooligan) [inline src/assignment-2/BasicVault.sol:38]: please research the [checks effects interactions pattern](https://docs.soliditylang.org/en/latest/security-considerations.html?highlight=checks%20effects%20interactions#use-the-checks-effects-interactions-pattern) and if you have any questions about it, please raise on slack for the benefit of everyone

Comment (devtooligan) [reply src/assignment-2/BasicVault.sol:38]: hmm i see u used it down below with withdraw where it is crucial.  it's probably ok to do it like this here...i'm just used to always sticking with that checks/effects/interactions pattern but maybe its not necessary here, @alcueca what do you think?

File: test/assignment-2/BasicVault.t.sol [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    ERC20MockWithFailedTransfers token;
+
+    function setUp() public virtual {
+        token = new ERC20MockWithFailedTransfers("Walkie Talkie", "WOKTOK");
+        v = new BasicVault(token);
+    }
+}
+
+contract ZeroStateTest is ZeroState {
+
```
Comment (devtooligan) [inline test/assignment-2/BasicVault.t.sol:19]: These tests are looking good in general.  It's clear you've got a good understanding of foundry testing and also the whole "state inheritance" pattern we talked about.

I have to shut down for the day but I will leave a general comment I have is it seems like you might have too many layers, making it overly complex and a bit hard to follow.  Not sure if you need the UserApproved and UserWithTokens layers and associated tests.  Also, the some of the tests you wrote are for the ERC20 code which, for this assignment, you can assume are fully tested.  I want to look at these more closely and give you better feedback, but that is my initial impression.  I like the `failTransfers`, thats nice!

File: src/assignment-2/BasicVault.sol (hunk 1 above)
Comment (alcueca) [reply src/assignment-2/BasicVault.sol:38]: When coding Yield v1 I did it this way, instead of using checks-effects-interactions. Basically you put the lines that are beneficial to the user last.

It is easier to understand, and that's why I did it. However, as complexity grows it becomes harder and harder to see what is safe and what is not. Checks-effects-interactions saves you from having to think about what actions are user-positive and in which order, and is safest in the long term.

Also, you will throw auditors off the scent if you don't use checks-effects-interactions, and even if you did it safely, they will spend extra time looking for issues there, instead of where you want them to be.

File: src/assignment-2/BasicVault.sol [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        balances[msg.sender] += amount;
+
+        emit Deposit(msg.sender, amount);
+    }
+
+    /// @notice Withdraws tokens back to the user
+    /// @dev Updates the balances mapping with the amount of tokens
+    /// @param amount The amount of tokens to be withdrawn
+    function withdraw(uint amount) public {
+        if (balances[msg.sender] < amount) {
```
Comment (alcueca) [inline src/assignment-2/BasicVault.sol:50]: One question, what would happen if you remove this conditional and error?

**Review by alcueca — COMMENTED** [review]
Some background, and a question.

File: src/assignment-2/BasicVault.sol (hunk 1 above)
Comment (davidbrai) [reply src/assignment-2/BasicVault.sol:38]: makes sense! thanks for the thoughtful feedback. I will change it to use the checks-effects-interactions (CEI ?) pattern.

File: src/assignment-2/BasicVault.sol (hunk 3 above)
Comment (davidbrai) [reply src/assignment-2/BasicVault.sol:50]: @alcueca good question :) I ran the tests, and it results in an underflow error.
I guess that for gas optimization this check can be removed.
is that what you had in mind?

File: src/assignment-2/BasicVault.sol (hunk 1 above)
Comment (davidbrai) [reply src/assignment-2/BasicVault.sol:38]: fixed here: https://github.com/davidbrai/yield-mentorship/pull/2/commits/fdc57876424aa9961a661227d7d231b4d9593f19

File: src/assignment-2/BasicVault.sol (hunk 3 above)
Comment (alcueca) [reply src/assignment-2/BasicVault.sol:50]: Yeah, that's exactly it. In this particular case the error check only gives a bit of information about what went wrong, which given that we can debug transactions with Tenderly, is not that relevant.

We could remove it and save some gas, or we could leave it and surround the subtraction afterwards in an `unchecked` block. Both options would be fine. For now, don't worry and move on.

**Comment by alcueca** [comment]
I don't think I can add myself as a reviewer, but please consider this PR approved.

**Comment by davidbrai** [comment]
Rinkeby deployment:
BasicVault: https://rinkeby.etherscan.io/address/0x9db9f594b9596b1751f29e3bc161b4b0b36d4ba3
token: https://rinkeby.etherscan.io/address/0x2129363821b37000d2c4e986d7ff438397866342
